### ISSUE TITLE
New SAML auth flow

### ISSUE THREAD

**Issue by msteffen** [pachyderm/pachyderm#3119]
This issue is revision of https://github.com/pachyderm/pachyderm/issues/3100
## Background

SAML support will require us to implement an auth flow similar to, but distinct from, the auth flow we have for oauth

## Design

Initially, to save time and effort, we will implement Okta-initiated authentication first (IdP-initiated in SAML terms)

##### Auth Flow for the Dash
- User configs dash address in auth config
- User gets form response from Okta (after logging in)
- User submits form to Pachyderm ACS endpoint
- Pachyderm logs user in, generates a temporary token, and returns a redirect to the dash with the temporary token in the Location URL
- Dash sends temporary token to Authenticate(), gets a real token, and stores that locally

##### Auth Flow for Pachctl
- User clicks a "Get temporary token" button in the dash
- Dash asks pachd for temp token (this API will be implemented as part of this project)
- Dash displays temp token
- User copy/pastes to pachctl
- Pachctl calls Authenticate with temporary token, receives a permanent token, and stores it

## Dash TODOs
- [ ] Todo: remove ability to modify groups
- [ ] Add ability to get a temporary token
- [ ] When displaying an ACL, have the ability to display group entries as well

## Pachctl/pachd TODOs
- [ ] **Blocking for Dash** Finish Pachd ACS service (recieves POST requests with a SAML response, validates it, and redirects the browser back to the dash with a temporary token)
  - The Dash will need this implemented for testing: this will be used to confirm that authentication works end-to-end, Okta to Dash
- [ ] **Blocking for Dash** Get temporary token API
  - The Dash will need this to implement the first step of the `pachctl` login flow above
- [ ] `Authenticate()`: Support adding groups to ACLs
- [ ] `Authorize()`: Support checking group memberships

**Comment by msteffen** [comment]
## Remaining work on this project:
I'm breaking up the rest of this work into three pieces for planning: 1) demo (what needs to be done so that we can show a working product and get credit for it), 2) finishing (what needs to be done before this is effectively off our plates and we can move on to other work, and 3) stretch goals that may be worth the marginal effort but should not block the project.

Also included are approximate time estimates for each section (not rigorous, but if/when we're past them, we should re-evaluate how our time is being spent and update them)

### Demo
2 business days—finished by EOD 9/10
##### Frontend
- [ ] Users can add groups to ACLs if they're not GitHub users or pipeline users
  - **Why**: The main consumer of Pachyderm's SAML and groups features will likely be doing ACL management in the dash and will definitely have groups on ACLs (perhaps exclusively) so they will need to be able to add groups to ACLs in the dash. However, adding groups to ACLs should be disabled for GitHub-only users, since there are not GitHub-only groups that they could add. This heuristic satisfies both constraints without requiring any more integration between the dash and pachd (for time's sake)
  - This is a hack—ultimately users should be able to add groups if the auth config includes a groups-enabled backend, but this is simpler and therefore faster
- [ ] Cosmetic fixes
  - OTP panel polls for a new OTP every second (but it should be static)
  - SAML users (and groups) have GitHub icons but shouldn't

##### Backend
- [ ] It should be possible to have groups in the set of cluster admins
  - **Why** The client will probably use groups for all of their access control (because they already have a group management system that they like), so they'll probably want to have a group (perhaps only a group) for the cluster admins that they manage externally, rather than having a de-facto group in Pachyderm that they have to manage through Pachyderm
- [ ] Respect SAML assertion expiration
  - **Why** This is particularly important because we're getting group membership info from SAML assertions. If a user is removed from a group, they won't lose access to the group's resources until they sign in again. It's expected that that will take as long as their last sign-in was valid for (something the client configures in their SAML ID provider) but unexpected that it would take longer than that.

### Finishing
##### Backend
- [ ] Finish writing tests for features in feature branch
  - **Why** we want to check this code into master, and we won't know it'll continue working there without tests
- [ ] Config validation
  - **Why** We should reject invalid configs, but right now we don't
- [ ] Config simplification
  - **Why** Users have to specify some information redundantly in their config. We should take this out of the config before anybody starts using it and we have to deal with config migration

### Stretch
##### Frontend
- [ ] Auto-complete in the ACL modification UI
  - **Why** I think this is a particularly valuable stretch goal—it's very easy to mistype a user's name and not grant them access to some protected resource (or worse, grant the wrong person access). By giving users auto-complete, we'll avoid a lot of confusion/complaints stemming from invalid ACLs.
- [ ] Display a Pachyderm cluster's list of users
  - **Why** Clients are paying per-user, so they should be able to see how they're using their budget
- [ ] Display a profile page for Pachyderm users when a user icon is clicked (rather than linking to a GitHub profile, which may not exist in the case of SAML users or robot users). Ideally this will also show the user's group memberships

##### Backend
APIs to support the above:
- [ ] Get the complete list of users in a cluster
- [ ] Get the complete list of groups in a cluster
- [ ] Get the auth config, or basic info from the auth config (particularly which ID backends are configured and whether they support groups)

**Comment by gabrielgrant** [comment]
One thing to discuss is what we want to do about multiple auth providers. My understanding is that right now github/OAuth-based auth is always active, even if saml is being used. The dash currently supports displaying users from multiple auth providers, but does not have the ability to add new users from an auth system other than one through which the current user is authenticated.

We should probably decide if multiple active auth systems is something we want to support (my gut feeling is probably not right now), and decide how urgent it is to forcefully deactivate oauth, or if that's something that can wait until we properly migrate auth-based auth to the new auth APIs (tracked in #3101)

**Comment by gabrielgrant** [comment]
Also, not specifically related to SAML, but possibly part of this project in general, I think we're supposed to change permissions around discover-ability (#2434), right?

**Comment by msteffen** [comment]
> One thing to discuss is what we want to do about multiple auth providers. My understanding is that right now github/OAuth-based auth is always active, even if saml is being used.
Correct, and I agree that changing this should be a long-term goal

> We should probably decide if multiple active auth systems is something we want to support (my gut feeling is probably not right now), and decide how urgent it is to forcefully deactivate oauth, or if that's something that can wait until we properly migrate auth-based auth to the new auth APIs/flows
I agree. Long-term: definite goal. Short-term: not needed yet.

**Comment by msteffen** [comment]
## Convo with Joey & JD

#### Conclusions:
- JD: doesn't make sense to hack a users table in the frontend before we have robust backend support. Shitty version of this isn't useful, so don't bother implementing it
- Gabriel should work on other stuff while SAML is being finished, then when SAML is done, come up with a good backend implementation of a users table, at which point displaying this info in the frontend will be Gabriel's priority.
- Would be useful, before SAML is done, to designing the dash view of the users table. Would help us figure out what information we'll want to have available to display.
  - Design should reflect what Joey already kind of knows what he wants (deactivating users? what does a user have access to?)

#### Notes
- Visibility of pipelines to certain users/groups, and tagging certain repos and pipelines as being related is below basic users table
- Visibility will, at some point, become urgent. But hasn't happened yet, and is probably a complicated feature, so postpone for now
- Users table is highest priority among auth things until we have another contract riding on a missing feature.
- How does that compare to other parts of Pachyderm? User profiles, for example, is less important. So is auto complete. The list above is in roughly the right order:
  - 1. List of users, 2. auto complete, 3. user profile page (i.e. admin wants to see what Alice has access to. When Alice gets fired, what do I have to revoke their access to), 4. OTP polling
- For now: if auth isn't activated, there's no way to know how many users are using the system. We're fine with that. We don't need to audit users who don't have auth activated because we don't think that's common
- Hackiest version of users table: call GetAcl on every repo and display a list of users on every ACL (doesn't work for groups though)
- Removing a user from every ACL they're part of

**Comment by gabrielgrant** [comment]
In talking with @msteffen we decided it doesn't really make sense to let users activate auth as a robot through the dashboard, so for testing I'll just do so directly through the API, then generate an OTP to put in the URL when redirecting to the otp-based autologin, and start the test from there.

**Comment by msteffen** [comment]
Many of the "stretch" goals previously on this issue have been moved to https://github.com/pachyderm/pachyderm/issues/3157, which is likely the next auth project
